pkg/controllers/gpu/controller_test.go

```diff
@@ -0,0 +1,181 @@
+			},
+		},
+		Spec: apps.DaemonSetSpec{
+			Selector: &metav1.LabelSelector{
+				MatchLabels: map[string]string{
+					"app": appLabel,
+				},
+			},
+			Template: v1.PodTemplateSpec{
+				ObjectMeta: metav1.ObjectMeta{
+					Labels: map[string]string{
+						"app": appLabel,
+					},
+				},
+				Spec: v1.PodSpec{
+					Containers: []v1.Container{
+						{
+							Name: appLabel,
+							Env: []v1.EnvVar{
+								{
+									Name: "NODE_NAME",
+									ValueFrom: &v1.EnvVarSource{
+										FieldRef: &v1.ObjectFieldSelector{
+											FieldPath: "spec.nodeName",
+										},
+									},
+								},
+							},
+							Args:            getPodArgs(devicePlugin),
+							Image:           devicePlugin.Spec.Image,
+							ImagePullPolicy: "IfNotPresent",
+							SecurityContext: &v1.SecurityContext{
+								ReadOnlyRootFilesystem: &yes,
+							},
+							VolumeMounts: []v1.VolumeMount{
+								{
+									Name:      "devfs",
+									MountPath: "/dev/dri",
+									ReadOnly:  true,
+								},
+								{
+									Name:      "sysfs",
+									MountPath: "/sys/class/drm",
+									ReadOnly:  true,
+								},
+								{
+									Name:      "kubeletsockets",
+									MountPath: "/var/lib/kubelet/device-plugins",
+								},
+							},
+						},
+					},
+					NodeSelector: nodeSelector,
+					Volumes: []v1.Volume{
+						{
+							Name: "devfs",
+							VolumeSource: v1.VolumeSource{
+								HostPath: &v1.HostPathVolumeSource{
+									Path: "/dev/dri",
+								},
+							},
+						},
+						{
+							Name: "sysfs",
+							VolumeSource: v1.VolumeSource{
+								HostPath: &v1.HostPathVolumeSource{
+									Path: "/sys/class/drm",
+								},
+							},
+						},
+						{
+							Name: "kubeletsockets",
+							VolumeSource: v1.VolumeSource{
+								HostPath: &v1.HostPathVolumeSource{
+									Path: "/var/lib/kubelet/device-plugins",
+								},
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+	// add the optional InitImage
+	if devicePlugin.Spec.InitImage != "" {
+		setInitContainer(&daemonSet.Spec.Template.Spec, devicePlugin.Spec.InitImage)
+	}
+	// add service account if resource manager is enabled
+	if devicePlugin.Spec.ResourceManager {
+		daemonSet.Spec.Template.Spec.ServiceAccountName = serviceAccountName
+		addVolumeIfMissing(&daemonSet.Spec.Template.Spec, "podresources", "/var/lib/kubelet/pod-resources", v1.HostPathDirectory)
+		addVolumeMountIfMissing(&daemonSet.Spec.Template.Spec, "podresources", "/var/lib/kubelet/pod-resources")
+	}
+	return &daemonSet
+}
+
+// Test that GPU daemonset created by using go:embed is
+// equal to the expected daemonset.
+func TestNewDaemonSetGPU(t *testing.T) {
+	c := &controller{}
+
+	pluginExpected := &devicepluginv1.GpuDevicePlugin{
+		Spec: devicepluginv1.GpuDevicePluginSpec{
+			InitImage:       "intel/intel-gpu-initcontainer:devel",
+			ResourceManager: true,
+		},
+	}
+	expected := c.newDaemonSetExpected(pluginExpected)
+
+	// use existing InitImage from the plugin yaml
```

Comment:
this needs thinking. if `NewDaemonSet()` always starts with an `InitImage` even when `Spec.InitImage` is not set, that not what we want. or if we move it so that InitImage is always used, we also need to change the `webhook.Defaulter` to set it if it's not set in the CR.

Reply:
@mythi this should be fixed now. I've also discovered quite nasty bug caused by not deepcopying global daemonset objects before modifying them. Added more test cases to gpu plugin to cover all possible variations of ResourceManager and InitImage related changes. PTAL.